src/kde/kde.cc

```diff
@@ -30,124 +30,437 @@ namespace rtt_kde {
+ */
+double kde::calc_cartesian_weight(const std::array<double, 3> &r0,
+                                  const std::array<double, 3> &one_over_h0,
+                                  const std::array<double, 3> &r,
+                                  const std::array<double, 3> &one_over_h,
+                                  const quick_index &qindex,
+                                  const double &discontinuity_cutoff) const {
+  double weight = 1.0;
+  for (size_t d = 0; d < qindex.dim; d++) {
+    const double u = (r0[d] - r[d]) * one_over_h0[d];
+    const double scale =
+        fabs(one_over_h0[d] - one_over_h[d]) / std::max(one_over_h0[d], one_over_h[d]) >
+                discontinuity_cutoff
+            ? 0.0
+            : 1.0;
+    // Apply Boundary Condition Weighting
+    double bc_weight = 1.0;
+    const bool low_reflect = reflect_boundary[d * 2];
+    const bool high_reflect = reflect_boundary[d * 2 + 1];
+    if (low_reflect) {
+      const double low_u =
+          ((r0[d] - qindex.bounding_box_min[d]) + (r[d] - qindex.bounding_box_min[d])) *
+          one_over_h0[d];
+      bc_weight += epan_kernel(low_u);
+    }
+    if (high_reflect) {
+      const double high_u =
+          ((qindex.bounding_box_max[d] - r0[d]) + (qindex.bounding_box_max[d] - r[d])) *
+          one_over_h0[d];
+      bc_weight += epan_kernel(high_u);
+    }
+    weight *= scale * bc_weight * epan_kernel(u) * one_over_h0[d];
+  }
+  return weight;
+}
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief calculate spherical weight
+ * 
+ * \pre Calculate the effective weight from a given location to the current kernel 
+ *
+ * \param[in] r0 current kernel center location
+ * \param[in] one_over_h0 current kernel width
+ * \param[in] r data location
+ * \param[in] one_over_h kernel width at this data location
+ * \param[in] qindex quick indexing class
+ * \param[in] discontinuity_cutoff maximum size of value discrepancies to include in the reconstruction
+ *
+ * \return weight contribution to the current kernel
+ *
+ * \post the local reconstruction of the original data is returned.
+ */
+double kde::calc_spherical_weight(const std::array<double, 3> &r0,
+                                  const std::array<double, 3> &one_over_h0,
+                                  const std::array<double, 3> &r,
+                                  const std::array<double, 3> &one_over_h,
+                                  const quick_index &qindex,
+                                  const double &discontinuity_cutoff) const {
+  // largest active smoothing length
+  const auto r0_theta_phi = qindex.transform_r_theta(sphere_center, r0);
+  // if we are near the origin of the sphere, fall back to xyz reconstruction
+  if (r0_theta_phi[0] < sphere_min_radius || r0_theta_phi[0] > sphere_max_radius)
+    return calc_cartesian_weight(r0, one_over_h0, r, one_over_h, qindex, discontinuity_cutoff);
```

Comment:
Codecov claims that this line is not covered by your tests.  If true, can you extend your tests to cover this case?

Reply:
Yap, I should be able to add this.